R/validate-policy.R

```diff
@@ -1,26 +1,35 @@
 #' Validate Policy File
 #' 
 #' Flattens and validates policy file
-#'  @param policy YAML policy file
-#'  @param gs Gradescope data
-#'  @importFrom purrr map discard
-#'  @export
+#' @param policy YAML policy file
+#' @param gs Gradescope data
+#' @importFrom purrr map discard
+#' @export
 validate_policy <- function(policy, gs){
   policy <- flatten_policy(policy)
-  # drop categories with unavailable assignments
-  categories <- map(policy$categories, "category") |> unlist()
-  # assignments is a vector that also includes category names
-  assignments <- c(get_assignments(gs), categories)
-  # this is because some categories have only nested categories as their assignments
-  # these categories should not be dropped
-  policy$categories <- map(policy$categories, function(cat){
-    cat$assignments <- cat$assignments[cat$assignments %in% assignments]
-    if (length(cat$assignments) == 0){
-      return (NULL)
-    }
-    return (cat)
-  }) |>
-    purrr::discard(is.null)
+  prev_length <- 0
+  current_length <- length(policy$categories)
+  #keep dropping until no more drops necessary
+  while (prev_length != current_length){
+    prev_length <- length(policy$categories)
+    categories <- map(policy$categories, "category") |> unlist()
+    assignments <- c(get_assignments(gs), categories)
+    policy$categories <- map(policy$categories, function(cat){
+      # drop categories with unavailable assignments/nested categories
+      cat$assignments <- cat$assignments[cat$assignments %in% assignments]
+      if (length(cat$assignments) == 0){
+        #if category has no assignments, dropp
+        return (NULL)
+      }
+      return (cat)
+    }) |>
+      purrr::discard(is.null)
+    current_length <- length(policy$categories)
+  }
+  
+  if (length(policy$categories) == 0){
+    return(NULL)
```

Comment:
Why are you returning `NULL` here? Would it make sense to throw an error message?

Reply:
I just needed a return statement to exit the function because the next section adds defaults, so there's no need to add defaults there's no valid policies. I'm not entirely sure about throwing an error because that depends on when we run validate_policy in the app... I thought we'd be running it as you're making categories, so it will throw errors even if you're just starting to make your categories. If there's a button that says "Calculate Grades", and only then do we validate the policy file, then yes, error makes sense

Reply:
When I wrote my version of this function, I was debating what should be done in this case. It could depend on where and how this function is called; however, for simplicity of error tracking it would be a good idea to launch any error from this function rather than letting NULL propagate through to other functions and causing errors since it would not be so obvious to a user why it is happening. I think it would be good to notify the user that their gradescope csv contains no assignments in the policy file since otherwise an instructor could be befuddled why Gradebook is not generating grades.

Reply:
There are three options to choose from here:

https://adv-r.hadley.nz/conditions.html#signalling-conditions

Something that we could do (this is what Zach is hinting at) is add an argument like `quiet` to each top-level function. Default could be set to `quiet = FALSE`, for running the package at the command line or in a script. Inside the app, we could call, if it made sense `quiet = TRUE`.  That's to say, we might want to route the signals differently depending on whether or not its run in the app.

Reply:
I put in the `quiet` argument for now... might have to change this if we're bring validate_policy() into get_grades(), but it'll work for running grades independent of the app